flocker/node/functional/test_deploy.py

```diff
@@ -116,64 +116,72 @@ def test_environment(self):
         """
         docker_dir = FilePath(__file__).sibling('env-docker')
         image = DockerImageBuilder(test=self, source_dir=docker_dir)
-        image_name = image.build()
 
-        application_name = random_name(self)
-
-        docker_client = DockerClient()
-        self.addCleanup(docker_client.remove, application_name)
-
-        volume_service = create_volume_service(self)
-        deployer = P2PNodeDeployer(
-            u"localhost", volume_service, docker_client,
-            make_memory_network(), node_uuid=uuid4())
-
-        expected_variables = frozenset({
-            'key1': 'value1',
-            'key2': 'value2',
-        }.items())
-
-        dataset = Dataset(
-            dataset_id=unicode(uuid4()),
-            metadata=pmap({"name": application_name}))
-        manifestation = Manifestation(dataset=dataset, primary=True)
-        desired_state = Deployment(nodes=frozenset([
-            Node(uuid=deployer.node_uuid,
-                 applications=frozenset([Application(
-                     name=application_name,
-                     image=DockerImage.from_string(
-                         image_name),
-                     environment=expected_variables,
-                     volume=AttachedVolume(
-                         manifestation=manifestation,
-                         mountpoint=FilePath('/data'),
-                         ),
-                     links=frozenset(),
-                     )]),
-                 manifestations={manifestation.dataset_id: manifestation})]))
-
-        d = change_node_state(deployer, desired_state)
-        d.addCallback(lambda _: volume_service.enumerate())
-        d.addCallback(lambda volumes:
-                      list(volumes)[0].get_filesystem().get_path().child(
-                          b'env'))
-
-        def got_result_path(result_path):
-            d = loop_until(result_path.exists)
-            d.addCallback(lambda _: result_path)
+        # TODO: convert this into a callback AZUL
+        outer_d = image.build()
+
+        # TODO: FLOC-2513
+        # This is horrible, but it's the simplest transformation to got
+        # from build() being pure to being monadic.
```

Comment:
I guess by "horrible" you mean that there should only be one chain of deferreds rather than an inner and outer deferred?  Add a new issue to clean this up, labelled `debt`, and refer to it here (and in the other places with the same comment) rather than referring to the issue about to be resolved.

(or clean it up).